external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSource.scala

```diff
@@ -0,0 +1,366 @@
+    val offset = KafkaSourceOffset(fetchPartitionOffsets(seekToLatest = true))
+    logDebug(s"GetOffset: ${offset.partitionToOffsets.toSeq.map(_.toString).sorted}")
+    Some(offset)
+  }
+
+  /** Returns the data that is between the offsets [`start`, `end`), i.e. end is exclusive. */
+  override def getBatch(start: Option[Offset], end: Offset): DataFrame = {
+    logInfo(s"GetBatch called with start = $start, end = $end")
+    val untilPartitionOffsets = KafkaSourceOffset.getPartitionOffsets(end)
+    val fromPartitionOffsets = start match {
+      case Some(prevBatchEndOffset) =>
+        KafkaSourceOffset.getPartitionOffsets(prevBatchEndOffset)
+      case None =>
+        initialPartitionOffsets
+    }
+
+    // Find the new partitions, and get their earliest offsets
+    val newPartitions = untilPartitionOffsets.keySet.diff(fromPartitionOffsets.keySet)
+    val newPartitionOffsets = if (newPartitions.nonEmpty) {
+      fetchNewPartitionEarliestOffsets(newPartitions.toSeq)
+    } else {
+      Map.empty[TopicPartition, Long]
+    }
+    if (newPartitionOffsets.keySet != newPartitions) {
+      // We cannot get from offsets for some partitions. It means they got deleted.
+      val deletedPartitions = newPartitions.diff(newPartitionOffsets.keySet)
+      logWarning(s"Partitions removed: ${deletedPartitions}, some data may have been missed")
+    }
+    logInfo(s"Partitions added: $newPartitionOffsets")
+    newPartitionOffsets.filter(_._2 != 0).foreach { case (p, o) =>
+      logWarning(s"Added partition $p starts from $o instead of 0, some data may have been missed")
+    }
+
+    val deletedPartitions = fromPartitionOffsets.keySet.diff(untilPartitionOffsets.keySet)
+    logWarning(s"Partitions removed: $deletedPartitions, some data may have been missed")
+
+    // Sort the partitions and current list of executors to consistently assign each partition
+    // to the executor. This allows cached KafkaConsumers in the executors to be re-used to
+    // read the same partition in every batch.
+    val topicPartitionOrdering = new Ordering[TopicPartition] {
+      override def compare(l: TopicPartition, r: TopicPartition): Int = {
+        implicitly[Ordering[(String, Long)]].compare(
+          (l.topic, l.partition),
+          (r.topic, r.partition))
+      }
+    }
+
+    // Use the until partitions to calculate offset ranges to ignore partitions that have
+    // been deleted
+    val sortedTopicPartitions = untilPartitionOffsets.keySet.filter { tp =>
+      newPartitionOffsets.contains(tp) || fromPartitionOffsets.contains(tp)
+    }.toSeq.sorted(topicPartitionOrdering)
+    logDebug("Sorted topicPartitions: " + sortedTopicPartitions.mkString(", "))
+
+    val sortedExecutors = getSortedExecutorList(sc)
+    val numExecutors = sortedExecutors.length
+    logDebug("Sorted executors: " + sortedExecutors.mkString(", "))
+
+    // Calculate offset ranges
+    val offsetRanges = sortedTopicPartitions.map { tp =>
```

Comment:
I'm lost on what the point of sorting topic partitions is, if you're still just using topicpartition hashcode as an index into the executors.